numpy/_core/defchararray.py

```diff
@@ -607,8 +607,12 @@ def decode(a, encoding=None, errors=None):
     array(['aAaAaA', '  aA  ', 'abBABba'], dtype='<U7')
 
     """
-    return _to_bytes_or_str_array(
-        _vec_string(a, object_, 'decode', _clean_args(encoding, errors)))
+    decoded_arr = _to_bytes_or_str_array(
+                             _vec_string(a, object_, 'decode',
+                             _clean_args(encoding, errors)))
+    if not (numpy.issubdtype(decoded_arr.dtype, numpy.str_)):
+        return decoded_arr.astype(numpy.str_)
```

Comment:
The `_to_bytes_or_str_array` function is weird.  The second argument should be required, but it isn't always used and the way it is used is always dubious.  It is either so old that there was logic missing that has been rewritten 2-3 times since then or just bugged.

I would suggest to refactor it to pass an explicit DType (without a length).  But easier than that is to just stop using that function.  You know the result must be unicode,/`np.str_`/`np.dtypes.StrDType`.

This is doing something wrong first and then trying fixing it, let's get it right first.  Also, please audit the other calls to `_to_bytes_or_str_array`.  There is at least one more where the second argument isn't even passed and it is in the same boat.
(I would like to change the ones which pass the second arg also, but that goes a bit further.)